### ISSUE TITLE
Calva output/REPL file conflict with two Calva sessions in separate VS Code windows for the same workspace root

### ISSUE THREAD

**Issue by gmp26** [BetterThanTomorrow/calva#1313]
Can we make the file path for output.calva-repl configurable to cope with this use case?

I’m running a Calva deps clj project in one VSCode workspace, and a Calva shadow-cljs project in another. They have nicely separate clj and cljs repls up and running except that they appear to share a common output.calva-repl. I get prompts like this on every cljs repl command. Is there any way to avoid this?
![image](https://user-images.githubusercontent.com/141701/135843897-b99c1830-2018-429a-bee8-fb0a4a8fa8b5.png)

**Comment by bpringe** [comment]
So, you are using one VS Code window, with two workspaces, and those two workspaces are sharing an output window? So each workspace has the same project directory in it, but you're using the workspaces like so to have a dedicated Clojure repl and a dedicated ClojureScript repl for the same project?

**Comment by gmp26** [comment]
I was following the One Folder Two Windows suggestion in https://calva.io/workspace-layouts/. So each workspace has its own window. And yes, the clojure and clojurescript repls are then independent except that they share the same output.calva-repl.

**Comment by bpringe** [comment]
Oh I see. @PEZ Have you seen this happen before?

I can't say I've used that workflow before, but  I haven't heard of this being an issue for others that I can remember. I do recall seeing that error before though, at least back when the output.calva-repl window was fairly new. That may have been due to us not saving it on writing though, before we started doing that.

Edit: I see this was discussed on Slack.

**Comment by gmp26** [comment]
There's a workaround for this problem. 

It only happens if you start the clj-repl BEFORE you start the cljs-repl.

Start the cljs-repl first and Robert is your uncle.

**Comment by PEZ** [comment]
> and Robert is your uncle

TIL. Such a wonderful expression!

**Comment by PEZ** [comment]
@bpringe, I don't understand why this happens, but I think it makes sense to be able to specify or nudge the file location somehow.

**Comment by PEZ** [comment]
Of course, if we can find the root cause and if it is within our reach to fix that, we should do that instead.

**Comment by bpringe** [comment]
> Robert is your uncle

@gmp26 is this a reference to Robert Martin, a.k.a. Uncle Bob? 🤔  😄

**Comment by gmp26** [comment]
@bpringe lol - No it’s an expansion of the English idiom ‘Bob’s your uncle’ meaning something like ‘and that fixes it’. I have no idea of it’s origin, but it predates even Uncle Bob.

**Comment by bpringe** [comment]
TIL 😃

**Comment by davidyang** [comment]
Hi @PEZ - thanks for Calva - really enjoying using it with Clojure.

I'm still seeing this problem even when loading the Shadow-CLJS first in one workspace and the deps.edn workspace in another.  Perhaps the easiest way to do this would be to allow overriding REPL_FILE_EXT per workspace? (or adding something like the environment or port number to the output file e.g. output-cljs.calva-repl/output-31412.calva-repl).

The evaluation works (so the Workspace is sending the code to the right repl) it's just that both windows stream each other's REPL buffers).

**Comment by PEZ** [comment]
Hi @davidyang, you are entirely welcome!

As I doubt we can do much about the root cause of this problem, I am thinking we will need to separate the output files in some nice way.

We can't really use the port, because that will leave tons of files open for many users and it will cause confusion about where the output is written. We could use some slug of the connect sequence. Maybe as a sub-directory:

```
.calva/output-window/shadow-cljs/output.calva.repl
```

And if you for some reason have multiple shadow-cljs REPLs sharing project root, in different windows, you create [custom connect sequences](https://calva.io/connect-sequences/) for them, that could only differ in their name. Say:

* My project A
* My project B

Then connecting using *My Project A* would create its file in:

```
.calva/output-window/my-project-a/output.calva.repl
```

That would keep it largely working as it does today for most users. For both you, @davidyang , and for @gmp26, it would start working w(o the write conflicts. And for the odd cases with same project type, there is a solution with configuring a connect sequence.
